pkg/kubelet/kuberuntime/fake_kuberuntime.go

```diff
@@ -0,0 +1,480 @@
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "PullImage")
+
+	// ImageID should be randomized for real container runtime, but here just use
+	// image's name for easily making fake images.
+	imageID := image.GetImage()
+	if _, ok := r.Images[imageID]; !ok {
+		r.Images[imageID] = makeFakeImage(image.GetImage())
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) RemoveImage(image *runtimeApi.ImageSpec) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "RemoveImage")
+
+	if _, ok := r.Images[image.GetImage()]; ok {
+		delete(r.Images, image.GetImage())
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) Version(apiVersion string) (*runtimeApi.VersionResponse, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "Version")
+
+	return &runtimeApi.VersionResponse{
+		Version:           &version,
+		RuntimeName:       &fakeRuntimeName,
+		RuntimeVersion:    &version,
+		RuntimeApiVersion: &version,
+	}, nil
+}
+
+func (r *fakeKubeRuntime) CreatePodSandbox(config *runtimeApi.PodSandboxConfig) (string, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "CreatePodSandbox")
+
+	// PodSandboxID should be randomized for real container runtime, but here just use
+	// sandbox's name for easily making fake sandboxes.
+	podSandboxID := config.GetName()
+	r.Sandboxes[podSandboxID] = &PodSandboxWithState{
+		config:    config,
+		state:     runtimeApi.PodSandBoxState_READY,
+		createdAt: time.Now().Unix(),
+	}
+
+	return podSandboxID, nil
+}
+
+func (r *fakeKubeRuntime) StopPodSandbox(podSandboxID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "StopPodSandbox")
+
+	if s, ok := r.Sandboxes[podSandboxID]; ok {
+		s.state = runtimeApi.PodSandBoxState_NOTREADY
+	} else {
+		return fmt.Errorf("pod sandbox %s not found", podSandboxID)
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) DeletePodSandbox(podSandboxID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "DeletePodSandbox")
+
+	if _, ok := r.Sandboxes[podSandboxID]; ok {
+		delete(r.Sandboxes, podSandboxID)
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) PodSandboxStatus(podSandboxID string) (*runtimeApi.PodSandboxStatus, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "PodSandboxStatus")
+
+	s, ok := r.Sandboxes[podSandboxID]
+	if !ok {
+		return nil, fmt.Errorf("pod sandbox %s not found", podSandboxID)
+	}
+
+	return &runtimeApi.PodSandboxStatus{
+		Id:        &podSandboxID,
+		Name:      s.config.Name,
+		CreatedAt: &s.createdAt,
+		State:     &s.state,
+		Network: &runtimeApi.PodSandboxNetworkStatus{
+			Ip: &fakePodSandboxIP,
+		},
+		Labels:      s.config.Labels,
+		Annotations: s.config.Annotations,
+	}, nil
+}
+
+func filterInLabels(filter, labels map[string]string) bool {
+	for k, v := range filter {
+		if value, ok := labels[k]; ok {
+			if value != v {
```

Comment:
Here can't just return because filter is also a map, other keys should be continue evaluated.